# LRTK
This is the genomic resequencing pipeline for read cloud data.
## Requirements
LRTK was written by python 3, and required packages are: os, sys, gzip, getopt, time, re, subprocess, random, string, glob, collections, multiprocessing, pysam

## Basic usage
python LRTK.py \<command\> [options]

### Command:
    Config		Generate configuration file
    Basicall	Execute the whole pipeline of basic data preparation
    Reseqall	Execute the whole pipeline of resequencing
    Denovoall	Execute the whole pipeline of de novo assembly
    Clean		delete temporary files

    Basic		Execute selected steps for basic data preparation
    Reseq		Execute selected steps for resequencing
    Denovo		Execute selected steps for de novo assembly
    
#### Config
    python LRTK_simple.py Config [options]
    Options:
    -o --outputdir, the path of output directory.
    -s --softwarepath, the path of directory where all software were installed [default: scriptdir/bin]
    -d --datasetpath, dataset directory [default: scriptdir/dataset]
    -b --bed, target region of WES/target sequencing data, or non-N region of the whole genome [default: scriptdir/dataset/genome/nonN.bed]. nonN.bed can be generated by find_nonN_region.py based on ‘scriptdir/dataset/genome/genome.fa’ automatically. File format: 1.chr id  2.Start  3.End 

#### Basicall
    python LRTK_simple.py Basicall [options]
    requisite options:
    -i --input, the input file containing fastq information (The input file contains three columns:1.Sample ID;2.Library ID;3. Path to sample fastqs).
    -o --outputdir, the path to output
    -c --config, configuration file [default: outdir/config/Basic.config]

    alternative options:
    -m --molecule, number. molecule length less than it would be discarded [default: 500]
    -p --parallel, the number of fq pairs that processing parallel. The max amount of invoking CPU would be 4*(-p) [default: 1]
    -N --noBX, generate additional fq file that has BX info or not [default: yes]

`Basicall` including `CFQ_ALN`, `MARK`, `BQSR`, `STAT` and `MERGE`, it would run all these patterns step by step. Moreover, you can also run these patterns step by step yourself, using `‘Basic’` patterns.

##### Basic:
>python LRTK_simple.py Basic \<command\> [options] <br>
>>`CFQ_ALN` generate clean fastq files and correct barcode error <br>
>>`MARK`    merge all bam files belong to the same library of each sample, and barcode aware PCR duplication removal (must complete ALN) <br>
>>`BQSR`    recalibrate base quality scores, using GATK. <br>
>>`STAT`    calculate QC statistics, including Cf, Cr, MuFL, NFP etc. (must complete CFQ and ALN) <br>
>>`MERGE`   merge all bam files belong to the same sample <br>

